### PR TITLE
[2817] - Location filter: live provider lookup

### PR DESCRIPTION
**Pull request by TobyRet**
### Context
Location filter page performs lookup when the option "By school, university or other training provider" is selected.

### Guidance to review
Separate ticket to be created for running JS tests on CI (note this needs to be done for Publish Teacher Training also). 

To run the JS tests, either:

1. `bundle exec rake` (also runs ruby specs and linters)
2. `yarn jest` (JS test only). 

auto_complete

### CODE REVIEW

**Comment by TobyRet** [comment]
[[M] location filter: live provider lookup (JS)](https://trello.com/c/hbegNDus/2817-m-location-filter-live-provider-lookup-js)

File: app/views/result_filters/location/new.html.erb [hunk 1]
```diff
@@ -111,13 +111,18 @@
...
-                  <span class="govuk-error-message" id="provider-error" data-qa="provider-error">
+                    <span class="govuk-error-message" id="provider-error" data-qa="provider-error">
                     <span class="govuk-visually-hidden">Error: </span><%= I18n.t("location_filter.errors.missing_provider") %>
                   </span>
                   <% end %>
                 <% end %>
                 <span class="govuk-hint">Enter the name or provider code</span>
-                <%= form.text_field :query, value: params[:query], class: "govuk-input", data: {qa: "provider-search"} %>
+                <%= form.text_field :query,
+                                    # autocomplete: 'off',
```
Comment (aliuk2012) [inline app/views/result_filters/location/new.html.erb]: should this be commented out?

File: app/webpacker/scripts/autocomplete.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  return (query, callback) => {
+    if (xhr && xhr.readyState !== XMLHttpRequest.DONE) {
+      xhr.abort();
+    }
+    const path = `${endpoint}?query=${query}`;
+
+    xhr = new XMLHttpRequest();
+    xhr.addEventListener("load", evt => {
+      let results = [];
+      try {
```
[the remark this answers is not in the document]
Comment (aliuk2012) [reply app/webpacker/scripts/autocomplete.js]: We could log those in Sentry, if we don't have a means of logging js errors.

File: app/webpacker/packs/application.js [hunk 3]
```diff
@@ -1,11 +1,23 @@
...
 window.initLocationsMap = initLocationsMap;
 
 const $backLink = document.querySelector('[data-module="back-link"]');
 new backLink($backLink).init();
+
+try {
+  const $autocomplete = document.getElementById("provider-autocomplete");
+  const $input = document.getElementById("query");
+  if ($autocomplete) {
+    initAutocomplete($autocomplete, $input);
```
Comment (defong) [inline app/webpacker/packs/application.js]: Would of thought this file would only contain just a call to initAutocomplete.

And initAutocomplete will contain the try catch along with the init code etc.

File: app/webpacker/packs/application.js [hunk 4]
```diff
@@ -1,11 +1,23 @@
...
 
 window.initLocationsMap = initLocationsMap;
 
 const $backLink = document.querySelector('[data-module="back-link"]');
 new backLink($backLink).init();
+
+try {
+  const $autocomplete = document.getElementById("provider-autocomplete");
+  const $input = document.getElementById("query");
+  if ($autocomplete) {
```
Comment (aliuk2012) [inline app/webpacker/packs/application.js]: Seeing as initAutoComplete needs `$input`, I think should also check that it exists before trying to call initAutoComplete. 

I do agree with @defong, I think this block should be part of initAutoComplete to keep `application.js` clean and simple to read.

File: app/views/result_filters/location/new.html.erb (hunk 1 above)
Comment (TobyRet) [reply app/views/result_filters/location/new.html.erb]: No. My bad

File: app/webpacker/packs/application.js (hunk 4 above)
Comment (TobyRet) [reply app/webpacker/packs/application.js]: 👍

File: app/webpacker/scripts/autocomplete.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/webpacker/scripts/autocomplete.js]: This autocomplete is calling a local route

Error should be logged via the inherited JSON API client https://github.com/DFE-Digital/find-teacher-training/blob/91daec19a3107863723d1364c2fe92f6b3b5b844/app/models/provider_suggestion.rb

File: app/webpacker/packs/application.js (hunk 4 above)
Comment (TobyRet) [reply app/webpacker/packs/application.js]: 👍

File: app/webpacker/packs/application.js (hunk 3 above)
Comment (TobyRet) [reply app/webpacker/packs/application.js]: @defong - looks like you have some refactoring to do!
https://github.com/DFE-Digital/search-and-compare-ui/blob/5cbef895923a452c7f20914e9c4acadba597f6eb/src/Assets/app.js#L40

Comment (defong) [reply app/webpacker/packs/application.js]: you copying bugs/defects again

**Comment by defong** [comment]
so is can the page progress after doing a search?

I get stuck 
image

https://s121d02-2817-find-as.azurewebsites.net/results/filter/location?qualifications=QtsOnly,PgdePgceWithQts,Other&fulltime=False&parttime=False&hasvacancies=True&senCourses=False

**Comment by TobyRet** [comment]
> so can the page progress after doing a search?
> 
> I get stuck
> image
> 
> https://s121d02-2817-find-as.azurewebsites.net/results/filter/location?qualifications=QtsOnly,PgdePgceWithQts,Other&fulltime=False&parttime=False&hasvacancies=True&senCourses=False

Had a JS typo in the autocomplete - fixed now

**Pull request merged**